### PR TITLE
Update from Facebook

### PR DESCRIPTION
**Pull request by orionr**

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -7707,7 +7707,8 @@ def load_tests(loader, tests, pattern):
         test_suite = unittest.TestSuite()
         for test_group in tests:
             for test in test_group:
-                if int(hashlib.sha256(str(test).encode('utf-8')).hexdigest(), 16) % num_shards == shard:
+                hash_id = int(hashlib.sha256(str(test).encode('utf-8')).hexdigest(), 16)
```
Comment (soumith) [inline test/test_nn.py]: approved

File: test/test_torch.py [hunk 2]
```diff
@@ -6668,10 +6668,7 @@ def import_module(name, filename):
             return module
 
         with filecontext_lambda() as checkpoint:
-            try:
-                fname = get_file_path_2(os.path.dirname(__file__), 'data', 'network1.py')
-            except IOError:
-                fname = get_file_path_2(os.path.dirname(__file__), 'data', 'network1.pyc')
+            fname = get_file_path_2(os.path.dirname(__file__), 'data', 'network1.py')
```
Comment (soumith) [inline test/test_torch.py]: approved

File: test/test_torch.py [hunk 3]
```diff
@@ -6684,10 +6681,7 @@ def import_module(name, filename):
                     self.assertEquals(len(w), 0)
 
             # Replace the module with different source
-            try:
-                fname = get_file_path_2(os.path.dirname(__file__), 'data', 'network2.py')
-            except IOError:
-                fname = get_file_path_2(os.path.dirname(__file__), 'data', 'network2.pyc')
+            fname = get_file_path_2(os.path.dirname(__file__), 'data', 'network2.py')
```
Comment (soumith) [inline test/test_torch.py]: approved

File: torch/_thnn/utils.py [hunk 4]
```diff
@@ -2,7 +2,6 @@
 import itertools
 import importlib
 
-# in fbcode, this fails in some cases, but we don't need it, therefore the try-catch
```
Comment (soumith) [inline torch/_thnn/utils.py]: approved

File: torch/csrc/autograd/anomaly_mode.h [hunk 5]
```diff
@@ -16,6 +16,7 @@ struct AnomalyMode {
 
 
 struct AnomalyMetadata {
+  virtual ~AnomalyMetadata(){};
```
Comment (soumith) [inline torch/csrc/autograd/anomaly_mode.h]: approved

File: torch/csrc/generic/StorageSharing.cpp [hunk 6]
```diff
@@ -101,10 +101,10 @@ static PyObject * THPStorage_(shareFilename)(THPStorage *self)
     ctx = (libshm_context*)allocator_obj->allocatorContext;
   } else {
     // TODO: retry on collision
-    AutoNoGIL no_gil;
+    // TODO: free GIL - but remember to reacquire it when an exception is thrown
     THWStoragePtr new_storage(THPStorage_(newFilenameStorage)(storage->size));
-    THWStorage_(copy)(new_storage, storage);
```
Comment (soumith) [inline torch/csrc/generic/StorageSharing.cpp]: looks incorrect, prob revert changes to this file.

File: torch/csrc/generic/StorageSharing.cpp [hunk 7]
```diff
@@ -182,10 +182,9 @@ static PyObject * THPStorage_(shareFd)(THPStorage *self)
     auto allocator_obj = ((StorageWeakRefAllocator*)storage->allocatorContext);
     ctx = (THMapAllocatorContext*)allocator_obj->allocatorContext;
   } else {
-    AutoNoGIL no_gil;
     THWStoragePtr new_storage(THPStorage_(newFdStorage)(storage->size));
-    THWStorage_(copy)(new_storage, storage);
```
Comment (soumith) [inline torch/csrc/generic/StorageSharing.cpp]: looks incorrect, prob revert changes to this file.

File: torch/csrc/jit/script/compiler.cpp [hunk 8]
```diff
@@ -316,16 +316,18 @@ void liftConstantAttributes(const FunctionSchema& schema, Node* node) {
...
     // this was a builtin with a vararg list lowered,
-    if(arg.type->kind() == TypeKind::ListType) {
+    if (arg.type->kind() == TypeKind::ListType) {
       // we do not support constant lifting of the arg itself
-      if(arg.attribute_info)
+      if (arg.attribute_info)
         return;
-      // but we do support it for other values so we need to skip all the vararg nodes:
-      size_t vararg_list_size = node->inputs().size() - (schema.arguments.size() - 1);
-      while(n < i + vararg_list_size) {
```
Comment (soumith) [inline torch/csrc/jit/script/compiler.cpp]: looks like some kind of auto-formatting, revert changes to this file.

File: torch/csrc/onnx/onnx.npb.cpp [hunk 9]
```diff
@@ -144,7 +144,7 @@ PB_STATIC_ASSERT((pb_membersize(onnx_TensorProto, segment) < 65536), YOU_MUST_DE
 #if !defined(PB_FIELD_16BIT) && !defined(PB_FIELD_32BIT)
 /* If you get an error here, it means that you need to define PB_FIELD_16BIT
  * compile-time option. You can do that in pb.h or on compiler command line.
- * 
+ *
```
Comment (soumith) [inline torch/csrc/onnx/onnx.npb.cpp]: approved

File: torch/onnx/symbolic.py [hunk 10]
```diff
@@ -599,6 +599,10 @@ def index_select(g, self, index, dim):
     return g.op("Gather", self, index, axis_i=dim)
 
 
+def index_put(g, *inputs, **kwargs):
```
Comment (soumith) [inline torch/onnx/symbolic.py]: approved

**Review by soumith — APPROVED** [review]
approved `torch` changes, but some file changes need to be removed

File: torch/csrc/generic/StorageSharing.cpp (hunk 6 above)
Comment (orionr) [reply torch/csrc/generic/StorageSharing.cpp]: Actually, I think this was https://github.com/pytorch/pytorch/pull/8887/commits/0ee126778f388a2fb13a115a604d68c51f89c54f and an actual failure internally. Do you still want to skip these changes?

Comment (soumith) [reply torch/csrc/generic/StorageSharing.cpp]: the revert commit is wrong. It makes `THWStoragePtr` into `THStoragePtr` (which will result in a compile error, as you see)

**Review by ezyang — APPROVED** [review]

**Comment by orionr** [comment]
rocm passed before so merging.

**Pull request merged**